### ISSUE TITLE
 Calculation of some logarithms can be slightly optimized

### ISSUE THREAD

**Issue by 0xVolosnikov** [OpenZeppelin/openzeppelin-contracts#4501]
**📝 Details**

### **Redundant shifts in "ifs"**
`log2` and `log256` contain "ifs" which can be simplified:
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/5ae630684a0f57de400ef69499addab4c32ac8fb/contracts/utils/math/Math.sol#L270

such conditions may be replaced by the following:
```solidity
if (value >= 2**128) {
```
According to my check, such a replacement saves about 40 gas (~12%) in the `log2`. 
On the other hand, this may increase the bytecode of the contract due to the need to write large constants. Perhaps that is why the checks are implemented in their current form.

### **Redundant variable on stack**

Logarithm calculations contain a redundant variable:

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/5ae630684a0f57de400ef69499addab4c32ac8fb/contracts/utils/math/Math.sol#L268

If instead you use a named variable in the "returns" and remove the explicit return, then this can save on one or two opcodes (3-5 gas)

Named return value is used, for example, in `mulDiv`.

### **Can be done without branches**

In the steps of computing logarithms (`log2`, `log256`), you can get rid of logical branching. Example:
https://github.com/cryptoalgebra/Algebra/blob/f43ae88856799e694170d5700d23c1657757ee9b/src/core/contracts/libraries/TickMath.sol#L83-L123

For 0 and 1 it will most likely be more expensive. But for other values, i think, cheaper. At the same time, of course, such code in assembly is less readable.

**Comment by Amxx** [comment]
**About the first part**

Are you sure ? My measurement say otherwize. I believe it is because `gt` is an evm opcode, but gte isn't (so its a compiled to an `lt` + `not` with adds 3 gas). I'm pretty sure we tested the option you proposed and ended up rejecting it. Maybe the compiler evolved and its not more effective...

**Comment by Amxx** [comment]
**About the second part**

We are not fans of relying on implicit variable declaration through named return type. If possible we avoid this. Ok, it can save 3~5 gas, but I'm not sure that is a significant saving that we want to compromise our guidelines for.

**Comment by Amxx** [comment]
**About the last part**

We value readability and auditability of the code. What is done in the link you shared doesn't really align with our best practices. In particular, if an error what to be made, we would not notice it.

**Comment by vladyan18** [comment]
> Are you sure ? My measurement say otherwize. I believe it is because `gt` is an evm opcode, but gte isn't (so its a compiled to an `lt` + `not` with adds 3 gas). I'm pretty sure we tested the option you proposed and ended up rejecting it. Maybe the compiler evolved and its not more effective...

Yes, I know about opcodes. But lately I've noticed that "ifs" with `gte` are sometimes more efficient. Maybe more efficient optimizations and JUMPI opcode layout are used.
I specifically checked three options: 
- old `if (value >> 128 > 0) { `
- with gt `if (value > 2 ** 128 - 1) { ` (compiler will calculate constant)
- with gte `if (value >= 2 ** 128) {`
and i my results, gas is: gte < gt < old

Contract used for gas tests (0.8.20):
```solidity
contract TestLog2 {
    function log2(uint256 value) internal pure returns (uint256) {
        uint256 result = 0;
        unchecked {
            if (value >= 2 ** 128) {
                value >>= 128;
                result += 128;
            }
            if (value >=  2 ** 64) {
                value >>= 64;
                result += 64;
            }
            if (value >= 2 ** 32 ) {
                value >>= 32;
                result += 32;
            }
            if (value >= 2 ** 16) {
                value >>= 16;
                result += 16;
            }
            if (value >=  2 ** 8) {
                value >>= 8;
                result += 8;
            }
            if (value >=  2 ** 4) {
                value >>= 4;
                result += 4;
            }
            if (value >=  2 ** 2) {
                value >>= 2;
                result += 2;
            }
            if (value > 1) {
                result += 1;
            }
        }
        return result;
    }

    function log2Old(uint256 value) internal pure returns (uint256) {
        uint256 result = 0;
        unchecked {
            if (value >> 128 > 0) {
                value >>= 128;
                result += 128;
            }
            if (value >> 64 > 0) {
                value >>= 64;
                result += 64;
            }
            if (value >> 32 > 0) {
                value >>= 32;
                result += 32;
            }
            if (value >> 16 > 0) {
                value >>= 16;
                result += 16;
            }
            if (value >> 8 > 0) {
                value >>= 8;
                result += 8;
            }
            if (value >> 4 > 0) {
                value >>= 4;
                result += 4;
            }
            if (value >> 2 > 0) {
                value >>= 2;
                result += 2;
            }
            if (value >> 1 > 0) {
                result += 1;
            }
        }
        return result;
    }


    function gasCostLog(uint256 x) external view returns (uint256 res, uint256 gas) {
        gas = gasleft();
        res = log2(x);
        gas -= gasleft();
    }

        function gasCostLogOld(uint256 x) external view returns (uint256 res, uint256 gas) {
        gas = gasleft();
        res = log2Old(x);
        gas -= gasleft();
    }
}
```

**Comment by vladyan18** [comment]
The compiler reverses the condition (to `lt`) by swapping execution branches
(Remix, assembly in compilation details)
![image](https://github.com/OpenZeppelin/openzeppelin-contracts/assets/13525908/0247099b-053c-4f67-bc91-f3938a3b6e39)

**Comment by vladyan18** [comment]
So, result with "1025" as input (Optimizer 200 runs, 0.8.20):
old: 322
with gt: 298
with gte: 283

**Comment by james-toussaint** [comment]
Optimized in [v5.3.0](https://github.com/OpenZeppelin/openzeppelin-contracts/tree/v5.3.0): `log2` & `log256` rewritten branchless with lookup table in https://github.com/OpenZeppelin/openzeppelin-contracts/pull/5236 and https://github.com/OpenZeppelin/openzeppelin-contracts/pull/5284.
